### PR TITLE
Batch Event Endpoint

### PR DESCRIPTION
**Pull request by agentfll**
:hourglass: **Status**: _Open for visibility_
:tickets: **Ticket(s)**:

## :construction_worker: Changes
- Implement batch event endpoint
*Note:* this implementation uses naive geography logic and filters by location in memory. This will change in the future when the location storage format in the database changes.

## :flashlight: Testing Instructions
1. In your terminal: `docker-compose up`
2. In your browser:
    - `http://localhost:8080/events`
    - `http://localhost:8080/events?start_time=2017-10-10T12:00:00%2B09:00&end_time=2017-11-17T12:00:00%2B09:00`
    - `http://localhost:8080/events?lat=49&lon=-123&radius=50`

### CODE REVIEW

File: beluga/routes.py [hunk 1]
```diff
@@ -4,8 +4,12 @@
 from sanic.exceptions import abort
 from sanic.response import json
 
+from beluga.models import session_scope, Event
 from beluga.auth import authorized
 
+import json as pyjson
+from datetime import datetime as dt, timezone as tz
+import math
```
Comment (chadlagore) [inline beluga/routes.py]: [Nitpick]
Idiomatic import ordering: base packages, 3rd party packages, then local modules, all in alphabetical order. 
☝️ These 3 are base packages.

File: beluga/routes.py [hunk 2]
```diff
@@ -66,7 +70,84 @@
...
+    lat = request.args.get('lat')
+    lon = request.args.get('lon')
+    radius = request.args.get('radius')
+
+    # Validate the parameters we got, if any
+    if start_time or end_time:
+        if not start_time or not end_time:
+            raise abort(400, "temporal range must be fully specified")
+        
+    if lat or lon or radius:
```
Comment (chadlagore) [inline beluga/routes.py]: Same here.

File: beluga/routes.py [hunk 3]
```diff
@@ -66,7 +70,84 @@
...
+            # https://en.wikipedia.org/wiki/Latitude#Length_of_a_degree_of_latitude.
+            lat_factor = 1.0 / 111.0 # 1 deg lat ~= 111 km (give or take 1 km)
+
+            min_lon = float(lon) - (float(radius) * lon_factor)
+            max_lon = float(lon) + (float(radius) * lon_factor)
+
+            min_lat = float(lat) - (float(radius) * lat_factor)
+            max_lat = float(lat) + (float(radius) * lat_factor)
+
+            # TODO once we can query by location, the query
```
Comment (chadlagore) [inline beluga/routes.py]: :+1:

File: beluga/routes.py [hunk 4]
```diff
@@ -66,7 +70,84 @@
...
+            # should be modified here to reflect the location we 
+            # got from the client.
+            pass
+        events = events.order_by(Event.start_time.asc())
+
+    # Convert from DB tables to dictionaries
+    # (read: deal with the JSON string in location)
+    events = map(lambda event: {
+        'title': event.title,
+        'location': {k:v for (k,v) in
```
Comment (chadlagore) [inline beluga/routes.py]: 👏

File: tests/test_routes.py [hunk 5]
```diff
@@ -9,7 +9,7 @@ def test_healthcheck_returns_200():
 def test_event_handler_exists():
     route = '/events'
     request, response = app.test_client.get(route)
-    assert response.status == 501
+    assert response.status == 200
```
Comment (chadlagore) [inline tests/test_routes.py]: :+1:

File: beluga/routes.py [hunk 6]
```diff
@@ -66,7 +70,78 @@
...
+        'start_time': event.start_time.astimezone(tz.utc).isoformat(),
+        'end_time': event.end_time.astimezone(tz.utc).isoformat()
+    }, events)
+
+    # HACKHACK we're filtering events by location ENTIRELY IN-MEMORY
+    # here. This is ugly, but necessary because we can't filter by
+    # lat/lon in the database because of how location is currently
+    # stored. This will change in the future.
+    if lat:
+        events = filter(lambda event: event['location']['lat'] > min_lat and event['location']['lat'] < max_lat and
```
Comment (chadlagore) [inline beluga/routes.py]: A bit more readable:

```python
events = filter(
            lambda event: (
                max_lat > event['location']['lat'] > min_lat and
                max_lon > event['location']['lon'] > min_lon
            ), events
        )
```

File: beluga/routes.py [hunk 7]
```diff
@@ -4,8 +4,12 @@
 from sanic.exceptions import abort
 from sanic.response import json
 
+from beluga.models import session_scope, Event
 from beluga.auth import authorized
 
+import json as pyjson
+from datetime import timezone as tz
+import math
```
Comment (chadlagore) [inline beluga/routes.py]: [Nitpick] Idiomatic imports: alphabetic and using 3 blocks: standard libraries, 3rd party, local modules.
These three new ones are standard libraries, so put em at the top ☝️.

File: tests/test_routes.py (hunk 5 above)
Comment (chadlagore) [reply tests/test_routes.py]: Not sure why this test is breaking on travis :eyes:

File: beluga/routes.py [hunk 8]
```diff
@@ -66,7 +70,78 @@
...
+            events = events.filter(Event.end_time <= end_time)
+        if lat:
+            # Convert from lat/lon/radius to upper/lower bounds for lat/lon
+
+            # We need to find the length of one degree longitude at the
+            # center point we were given. To do so, we take the cosine
+            # of the latitude we are given and multiply that by the length
+            # of one degree longitude at the equator.
+            lat_rads = (math.pi * float(lat)) / 180.0 # given lat in radians
+            lon_factor = 1.0 / ((math.pi / 180.0) * 6371.0 * math.cos(lat_rads))
```
Comment (chadlagore) [inline beluga/routes.py]: Can we pull `6371.0` out into a variable?

**Comment by chadlagore** [comment]
Oh looks like passing on travis :+1:

File: beluga/routes.py [hunk 9]
```diff
@@ -66,7 +70,78 @@
...
+            min_lon = float(lon) - (float(radius) * lon_factor)
+            max_lon = float(lon) + (float(radius) * lon_factor)
+
+            min_lat = float(lat) - (float(radius) * lat_factor)
+            max_lat = float(lat) + (float(radius) * lat_factor)
+
+            # TODO once we can query by location, the query
+            # should be modified here to reflect the location we 
+            # got from the client.
+            pass
```
Comment (chadlagore) [inline beluga/routes.py]: :gun:

File: beluga/routes.py (hunk 6 above)
Comment (chadlagore) [reply beluga/routes.py]: (I know we'll kill this code later).

File: beluga/routes.py [hunk 10]
```diff
@@ -66,7 +70,78 @@
...
+                pyjson.loads(event.location).items()
+                if k in ['lat', 'lon', 'title']},
+        'start_time': event.start_time.astimezone(tz.utc).isoformat(),
+        'end_time': event.end_time.astimezone(tz.utc).isoformat()
+    }, events)
+
+    # HACKHACK we're filtering events by location ENTIRELY IN-MEMORY
+    # here. This is ugly, but necessary because we can't filter by
+    # lat/lon in the database because of how location is currently
+    # stored. This will change in the future.
```
Comment (chadlagore) [inline beluga/routes.py]: I'm not convinced this computation is taking place in memory. `event` is still a query at this point, you're still writing SQL as far as I can tell.

File: beluga/routes.py (hunk 7 above)
Comment (agentfll) [reply beluga/routes.py]: will do

File: beluga/routes.py (hunk 8 above)
Comment (agentfll) [reply beluga/routes.py]: yep, good call

File: beluga/routes.py (hunk 6 above)
Comment (agentfll) [reply beluga/routes.py]: will do. i was thinking about doing this and ended up going one line each for lat & long, but i think that it's too long as is.

File: beluga/routes.py (hunk 9 above)
Comment (agentfll) [reply beluga/routes.py]: i just found out about `pass` and now i'm enamored with it!!

File: beluga/routes.py (hunk 10 above)
Comment (agentfll) [reply beluga/routes.py]: my understanding is that the query is executed when we treat `events` as a collection (i.e. when we invoke `filter`)

File: beluga/routes.py (hunk 9 above)
Comment (chadlagore) [reply beluga/routes.py]: Im a fan as well, I don't think it's doing anything here though! Or am I mistaken?

Comment (agentfll) [reply beluga/routes.py]: i was using it as a placeholder for the query-by-location-in-the-database code

**Pull request merged**